app/ts-cli/geminiql/lexer.go

```diff
@@ -0,0 +1,326 @@
+		}
+	}
+
+	tok := t.scanKeywords(buf.String())
+	return tok, buf.String()
+}
+
+func (t *Tokenizer) scanKeywords(s string) int {
+	if tok, ok := t.keywords[strings.ToUpper(s)]; ok {
+		return tok
+	}
+	return IDENT
+}
+
+func (t *Tokenizer) scanRaw() (int, string) {
+	var buf bytes.Buffer
+	for {
+		ch := t.read()
+		if ch == ',' || ch == ' ' {
+			t.unRead()
+			return RAW, buf.String()
+		} else if ch == EOF {
+			return RAW, buf.String()
+		} else {
+			buf.WriteRune(ch)
+		}
+	}
+}
+
+func (t *Tokenizer) scanString() (int, string) {
+	end := t.read()
+
+	var buf bytes.Buffer
+	for {
+		ch := t.read()
+		if ch == end {
+			return STRING, buf.String()
+		} else if ch == EOF || ch == '\n' {
+			return BAD_STRING, buf.String()
+		} else if ch == '\\' {
+			nch := t.read()
+			if nch == 'n' {
+				buf.WriteRune('\n')
+			} else if nch == '\\' {
+				buf.WriteRune('\\')
+			} else if nch == '"' {
+				buf.WriteRune('"')
+			} else if nch == '\'' {
+				buf.WriteRune('\'')
+			} else {
+				return BAD_ESCAPE, string(ch) + string(nch)
+			}
+		} else {
+			buf.WriteRune(ch)
+		}
+	}
+}
+
+func (t *Tokenizer) scanDigit() (int, string) {
+	var buf bytes.Buffer
+
+	typ := INTEGER
+	for {
+		ch := t.read()
+
+		if ch == EOF {
+			break
+		} else if ch == '.' {
+			nch := t.Lookahead()
+			if !unicode.IsDigit(nch) {
+				t.unRead()
+				break
+			} else {
+				buf.WriteRune(ch)
+				typ = DECIMAL
+			}
+		} else if !unicode.IsDigit(ch) {
+			t.unRead()
+			break
+		} else {
+			buf.WriteRune(ch)
+		}
+	}
+	return typ, buf.String()
+}
+
+func (t *Tokenizer) scanWhiteSpace() (int, string) {
+	var buf bytes.Buffer
+
+	for {
+		ch := t.read()
+		if ch == EOF {
+			break
+		} else if !unicode.IsSpace(ch) {
+			t.unRead()
+			break
+		} else {
+			buf.WriteRune(ch)
+		}
+	}
+	return WS_TOKEN, buf.String()
+}
+
+func (t *Tokenizer) Lookahead() rune {
+	ch, _, err := t.r.ReadRune()
+	defer t.unRead()
+
+	if err != nil {
+		ch = EOF
+	}
+	return ch
+}
+
+func (t *Tokenizer) read() rune {
+	ch, _, err := t.r.ReadRune()
+
+	if err != nil {
+		ch = EOF
+	}
+	return ch
+}
+
+func (t *Tokenizer) unRead() error {
+	return t.r.UnreadRune()
+}
+
+type QLLexerImpl struct {
+	tokenizer *Tokenizer
+	ast       *QLAst
+}
+
+func QLNewLexer(tokenizer *Tokenizer, ast *QLAst) QLLexer {
+	l := &QLLexerImpl{
+		tokenizer: tokenizer,
+		ast:       ast,
+	}
+
+	return l
+}
+
+func (l *QLLexerImpl) Lex(lval *QLSymType) int {
+	var tok int
+	var val string
+
+	for {
+		tok, val = l.tokenizer.Scan()
+		switch tok {
+		case ILLEGAL_TOKEN:
+			l.Error(fmt.Sprintf("illegal token %s", val))
+		case EOF_TOKEN:
+			return 0
+		case BAD_STRING:
+			l.Error(fmt.Sprintf("bad string %s", val))
+		case BAD_ESCAPE:
+			l.Error(fmt.Sprintf("bad escape %s", val))
+		case INTEGER:
+			var err error
+			lval.integer, err = strconv.ParseInt(val, 10, 64)
+			if err != nil {
+				l.Error(err.Error())
+			}
+		case DECIMAL:
+			var err error
+			lval.decimal, err = strconv.ParseFloat(val, 64)
```

Comment:
1. DECIMAL is a SQL keyword and data type, which is different with FLOAT in SQL.
    It is also different with float32/64 in Golang.  Should be improved in future.
2. Scientific notation:
    Please confirm if scientific notation is supported in Influx and openGemini.
    Tests show it is supported in openGemini, but has some bugs.
>>> insert sensor,farmID="001",device="110" sensorId="s20",value=50.98e0
Elapsed: 1.0919ms
>>> insert sensor,farmID="001",device="110" sensorId="s20",value=50.98e999
Elapsed: 1.3676ms
>>>  select * from sensor where value!=50.98
Elapsed: 2.469ms
ERR: json: unsupported value: +Inf
3. Please also confirm if+Inf, -Inf, NaN are supported or not.

Reply:
1. The syntax parser in command line is used to parse the interactive commands. SQL parsing on the server is not affected now. But it should be improved in future.
2. The Line Protocol is parsed as raw data in command line. It will be completely parsed on the server side. It looks like there are bugs on server side when encoding +Inf number to json. It should be fixed later.
3. Got it, those will be confirmed.

Reply:
I will fire a bug on the scientific notion.